icelake/src/io/functional_writer/partition_writer.rs

```diff
@@ -0,0 +1,212 @@
+//! task_writer module provide a task writer for writing data in a table.
+//! table writer used directly by the compute engine.
+use crate::error::Result;
+use crate::io::IcebergWriteResult;
+use crate::io::IcebergWriter;
+use crate::io::IcebergWriterBuilder;
+use crate::types::Any;
+use crate::types::FieldProjector;
+use crate::types::PartitionKey;
+use crate::types::PartitionSpec;
+use crate::types::PartitionSplitter;
+use arrow_array::RecordBatch;
+use arrow_schema::SchemaRef;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+/// PartitionWriter can route the batch into different inner writer by partition key.
+#[derive(Clone)]
+pub struct PartitionedWriterBuilder<B: IcebergWriterBuilder> {
+    inner: B,
+    partition_type: Any,
+    partition_spec: PartitionSpec,
+}
+
+impl<B: IcebergWriterBuilder> PartitionedWriterBuilder<B> {
+    pub fn new(inner: B, partition_type: Any, partition_spec: PartitionSpec) -> Self {
+        Self {
+            inner,
+            partition_type,
+            partition_spec,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<B: IcebergWriterBuilder> IcebergWriterBuilder for PartitionedWriterBuilder<B>
+where
+    B::R: IcebergWriter,
+{
+    type R = PartitionedWriter<B>;
+
+    async fn build(self, schema: &SchemaRef) -> Result<Self::R> {
+        let (projector, _) =
+            FieldProjector::new(schema.fields(), &self.partition_spec.column_ids())?;
+        Ok(PartitionedWriter {
+            inner_writers: HashMap::new(),
+            partition_splitter: PartitionSplitter::try_new(
+                projector,
+                &self.partition_spec,
+                self.partition_type,
+            )?,
+            inner_buidler: self.inner,
+            schema: schema.clone(),
+        })
+    }
+}
+
+/// Partition append only writer
+pub struct PartitionedWriter<B: IcebergWriterBuilder>
```

Comment:
Should we call this `FanoutPartitionedWriter`?

Reply:
Yes. Nice suggestion. And we can introduce PartitionWriter in the future.